File: book/lib/services/pip_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_client.dart';

/// Service for managing Python packages via pip
class PipService {
  String get _baseUrl => ApiClient.baseUrl;

  /// Get list of installed packages
  Future<List<PipPackage>> listInstalled() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/api/packages'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final packages = data['packages'] as List<dynamic>? ?? [];
        return packages.map((p) => PipPackage.fromJson(p)).toList();
      }
    } catch (e) {
      print('Error listing packages: $e');
    }
    return [];
  }

  /// Install a package
  Future<PipInstallResult> install(String packageName, {String? version}) async {
    try {
      final body = {
        'package': version != null ? '$packageName==$version' : packageName,
      };

      final response = await http.post(
        Uri.parse('$_baseUrl/api/packages/install'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(body),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return PipInstallResult(
          success: data['success'] ?? false,
          message: data['message'] ?? '',
          output: data['output'] ?? '',
        );
      } else {
        final data = jsonDecode(response.body);
        return PipInstallResult(
          success: false,
          message: data['detail'] ?? 'Installation failed',
          output: '',
        );
      }
    } catch (e) {
      return PipInstallResult(
        success: false,
        message: 'Error: $e',
        output: '',
      );
    }
  }

  /// Uninstall a package
  Future<PipInstallResult> uninstall(String packageName) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/api/packages/uninstall'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'package': packageName}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return PipInstallResult(
          success: data['success'] ?? false,
          message: data['message'] ?? '',
          output: data['output'] ?? '',
        );
      } else {
        final data = jsonDecode(response.body);
        return PipInstallResult(
          success: false,
          message: data['detail'] ?? 'Uninstallation failed',
          output: '',
        );
      }
    } catch (e) {
      return PipInstallResult(
        success: false,
        message: 'Error: $e',
        output: '',
      );
    }
  }

  /// Search PyPI for packages
  Future<List<PyPIPackage>> search(String query) async {
    if (query.isEmpty) return [];

    try {
      // Use PyPI JSON API for search
      final response = await http.get(
        Uri.parse('https://pypi.org/pypi/$query/json'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final info = data['info'] as Map<String, dynamic>;
        return [
          PyPIPackage(
            name: info['name'] ?? query,
            version: info['version'] ?? '',
            summary: info['summary'] ?? '',
            author: info['author'] ?? '',
            homePage: info['home_page'] ?? '',
            license: info['license'] ?? '',
          ),
        ];
      }
    } catch (e) {
      print('Error searching PyPI: $e');
    }

    // If exact match fails, try to get suggestions
    return await _searchSuggestions(query);
  }

  /// Get package suggestions from PyPI
  Future<List<PyPIPackage>> _searchSuggestions(String query) async {
    try {
      // Use PyPI simple API to get matching packages
      final response = await http.get(
        Uri.parse('https://pypi.org/simple/'),
      );

      if (response.statusCode == 200) {
        final html = response.body;
        final regex = RegExp(r'>([^<]+)</a>', multiLine: true);
        final matches = regex.allMatches(html);

        final suggestions = <PyPIPackage>[];
        final queryLower = query.toLowerCase();

        for (final match in matches) {
          final name = match.group(1);
          if (name != null && name.toLowerCase().contains(queryLower)) {
            suggestions.add(PyPIPackage(
              name: name,
              version: '',
              summary: '',
              author: '',
              homePage: '',
              license: '',
            ));
            if (suggestions.length >= 20) break;
          }
        }

        return suggestions;
      }
    } catch (e) {
      print('Error getting suggestions: $e');
    }
    return [];
  }

  /// Get package details from PyPI
  Future<PyPIPackage?> getPackageInfo(String packageName) async {
    try {
      final response = await http.get(
        Uri.parse('https://pypi.org/pypi/$packageName/json'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final info = data['info'] as Map<String, dynamic>;
        final releases = data['releases'] as Map<String, dynamic>? ?? {};

        return PyPIPackage(
          name: info['name'] ?? packageName,
          version: info['version'] ?? '',
          summary: info['summary'] ?? '',
          author: info['author'] ?? '',
          homePage: info['home_page'] ?? info['project_url'] ?? '',
          license: info['license'] ?? '',
          description: info['description'] ?? '',
          requiresPython: info['requires_python'] ?? '',
          versions: releases.keys.toList().reversed.take(20).toList(),
        );
      }
    } catch (e) {
      print('Error getting package info: $e');
    }
    return null;
  }

  /// Get outdated packages
  Future<List<OutdatedPackage>> checkOutdated() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/api/packages/outdated'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final packages = data['packages'] as List<dynamic>? ?? [];
        return packages.map((p) => OutdatedPackage.fromJson(p)).toList();
      }
    } catch (e) {
      print('Error checking outdated: $e');
    }
    return [];
  }

  /// Upgrade a package
  Future<PipInstallResult> upgrade(String packageName) async {
    return install(packageName);
  }

  /// Export requirements.txt
  Future<String> exportRequirements() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/api/packages/requirements'),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['requirements'] ?? '';
      }
    } catch (e) {
      print('Error exporting requirements: $e');
    }
    return '';
  }

  /// Install from requirements.txt content
  Future<PipInstallResult> installFromRequirements(String requirements) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/api/packages/install-requirements'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'requirements': requirements}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return PipInstallResult(
          success: data['success'] ?? false,
          message: data['message'] ?? '',
          output: data['output'] ?? '',
        );
      }
    } catch (e) {
      return PipInstallResult(
        success: false,
        message: 'Error: $e',
        output: '',
      );
    }
    return PipInstallResult(
      success: false,
      message: 'Unknown error',
      output: '',
    );
  }
}

/// Installed pip package
class PipPackage {
  final String name;
  final String version;
  final String location;

  const PipPackage({
    required this.name,
    required this.version,
    this.location = '',
  });

  factory PipPackage.fromJson(Map<String, dynamic> json) {
    return PipPackage(
      name: json['name'] as String? ?? '',
      version: json['version'] as String? ?? '',
      location: json['location'] as String? ?? '',
    );
  }
}

/// PyPI package info
class PyPIPackage {
  final String name;
  final String version;
  final String summary;
  final String author;
  final String homePage;
  final String license;
  final String description;
  final String requiresPython;
  final List<String> versions;

  const PyPIPackage({
    required this.name,
    required this.version,
    required this.summary,
    required this.author,
    required this.homePage,
    required this.license,
    this.description = '',
    this.requiresPython = '',
    this.versions = const [],
  });
}

/// Pip install result
class PipInstallResult {
  final bool success;
  final String message;
  final String output;

  const PipInstallResult({
    required this.success,
    required this.message,
    required this.output,
  });
}

/// Outdated package info
class OutdatedPackage {
  final String name;
  final String currentVersion;
  final String latestVersion;

  const OutdatedPackage({
    required this.name,
    required this.currentVersion,
    required this.latestVersion,
  });

  factory OutdatedPackage.fromJson(Map<String, dynamic> json) {
    return OutdatedPackage(
      name: json['name'] as String? ?? '',
      currentVersion: json['current_version'] as String? ?? json['version'] as String? ?? '',
      latestVersion: json['latest_version'] as String? ?? json['latest'] as String? ?? '',
    );
  }
}

/// Global pip service instance
final pipService = PipService();

File: notebook/src/services/pipService.ts
// Pip Service - Python package management

import apiClient from "./apiClient";

export interface InstalledPackage {
  name: string;
  version: string;
  location: string;
  requires: string[];
  requiredBy: string[];
}

export interface PackageInfo {
  name: string;
  version: string;
  summary: string;
  author: string;
  authorEmail?: string;
  license: string;
  homePage?: string;
  projectUrl?: string;
  requiresPython?: string;
  keywords: string[];
  classifiers: string[];
}

export interface PackageSearchResult {
  name: string;
  version: string;
  summary: string;
  score?: number;
}

export interface InstallProgress {
  status: "downloading" | "installing" | "complete" | "error";
  package: string;
  progress: number;
  message: string;
}

export interface InstallResult {
  success: boolean;
  package: string;
  version?: string;
  message: string;
  duration: number;
}

export interface UninstallResult {
  success: boolean;
  package: string;
  message: string;
}

export interface RequirementsParseResult {
  packages: Array<{
    name: string;
    version?: string;
    extras?: string[];
    markers?: string;
  }>;
  errors: string[];
}

interface PackageResponse {
  name: string;
  version: string;
  location: string;
  requires: string[];
  required_by: string[];
}

interface PackageInfoResponse {
  name: string;
  version: string;
  summary: string;
  author: string;
  author_email?: string;
  license: string;
  home_page?: string;
  project_url?: string;
  requires_python?: string;
  keywords: string[];
  classifiers: string[];
}

class PipService {
  private parsePackage(data: PackageResponse): InstalledPackage {
    return {
      name: data.name,
      version: data.version,
      location: data.location,
      requires: data.requires,
      requiredBy: data.required_by,
    };
  }

  private parsePackageInfo(data: PackageInfoResponse): PackageInfo {
    return {
      name: data.name,
      version: data.version,
      summary: data.summary,
      author: data.author,
      authorEmail: data.author_email,
      license: data.license,
      homePage: data.home_page,
      projectUrl: data.project_url,
      requiresPython: data.requires_python,
      keywords: data.keywords,
      classifiers: data.classifiers,
    };
  }

  // List installed packages
  async listPackages(): Promise<InstalledPackage[]> {
    const response = await apiClient.get<PackageResponse[]>("/api/pip/packages");
    return response.map((p) => this.parsePackage(p));
  }

  // Get package info
  async getPackageInfo(name: string): Promise<PackageInfo> {
    const response = await apiClient.get<PackageInfoResponse>(
      `/api/pip/packages/${encodeURIComponent(name)}`
    );
    return this.parsePackageInfo(response);
  }

  // Search packages on PyPI
  async searchPackages(query: string, limit: number = 20): Promise<PackageSearchResult[]> {
    const response = await apiClient.get<PackageSearchResult[]>(
      `/api/pip/search?query=${encodeURIComponent(query)}&limit=${limit}`
    );
    return response;
  }

  // Install package
  async installPackage(
    name: string,
    version?: string,
    upgrade: boolean = false
  ): Promise<InstallResult> {
    const response = await apiClient.post<{
      success: boolean;
      package: string;
      version?: string;
      message: string;
      duration: number;
    }>("/api/pip/install", {
      package: name,
      version,
      upgrade,
    });

    return response;
  }

  // Install multiple packages
  async installPackages(
    packages: Array<{ name: string; version?: string }>
  ): Promise<InstallResult[]> {
    const response = await apiClient.post<InstallResult[]>("/api/pip/install-multiple", {
      packages,
    });
    return response;
  }

  // Install from requirements.txt
  async installFromRequirements(content: string): Promise<InstallResult[]> {
    const response = await apiClient.post<InstallResult[]>("/api/pip/install-requirements", {
      content,
    });
    return response;
  }

  // Install with streaming progress
  async *installWithProgress(name: string, version?: string): AsyncGenerator<InstallProgress> {
    const stream = apiClient.streamSSE("/api/pip/install/stream", {
      package: name,
      version,
    });

    for await (const chunk of stream) {
      try {
        const data = JSON.parse(chunk);
        yield {
          status: data.status,
          package: data.package,
          progress: data.progress,
          message: data.message,
        };
      } catch {
        // Skip non-JSON chunks
      }
    }
  }

  // Uninstall package
  async uninstallPackage(name: string): Promise<UninstallResult> {
    const response = await apiClient.post<UninstallResult>("/api/pip/uninstall", {
      package: name,
    });
    return response;
  }

  // Uninstall multiple packages
  async uninstallPackages(names: string[]): Promise<UninstallResult[]> {
    const response = await apiClient.post<UninstallResult[]>("/api/pip/uninstall-multiple", {
      packages: names,
    });
    return response;
  }

  // Upgrade package
  async upgradePackage(name: string): Promise<InstallResult> {
    return this.installPackage(name, undefined, true);
  }

  // Check for outdated packages
  async checkOutdated(): Promise<
    Array<{
      name: string;
      currentVersion: string;
      latestVersion: string;
    }>
  > {
    const response = await apiClient.get<
      Array<{
        name: string;
        current_version: string;
        latest_version: string;
      }>
    >("/api/pip/outdated");

    return response.map((p) => ({
      name: p.name,
      currentVersion: p.current_version,
      latestVersion: p.latest_version,
    }));
  }

  // Generate requirements.txt
  async generateRequirements(): Promise<string> {
    const response = await apiClient.get<{ content: string }>("/api/pip/requirements");
    return response.content;
  }

  // Parse requirements.txt content
  async parseRequirements(content: string): Promise<RequirementsParseResult> {
    const response = await apiClient.post<{
      packages: Array<{
        name: string;
        version?: string;
        extras?: string[];
        markers?: string;
      }>;
      errors: string[];
    }>("/api/pip/parse-requirements", { content });

    return response;
  }

  // Check if package is installed
  async isInstalled(name: string): Promise<boolean> {
    try {
      await this.getPackageInfo(name);
      return true;
    } catch {
      return false;
    }
  }

  // Get package dependencies tree
  async getDependencyTree(name: string): Promise<{
    name: string;
    version: string;
    dependencies: Array<{
      name: string;
      version: string;
      required: string;
    }>;
  }> {
    const response = await apiClient.get<{
      name: string;
      version: string;
      dependencies: Array<{
        name: string;
        version: string;
        required: string;
      }>;
    }>(`/api/pip/packages/${encodeURIComponent(name)}/dependencies`);

    return response;
  }

  // Get PyPI package info (without installing)
  async getPyPIInfo(name: string): Promise<{
    name: string;
    version: string;
    summary: string;
    author: string;
    license: string;
    homePage?: string;
    projectUrls: Record<string, string>;
    releases: string[];
  }> {
    const response = await apiClient.get<{
      name: string;
      version: string;
      summary: string;
      author: string;
      license: string;
      home_page?: string;
      project_urls: Record<string, string>;
      releases: string[];
    }>(`/api/pip/pypi/${encodeURIComponent(name)}`);

    return {
      name: response.name,
      version: response.version,
      summary: response.summary,
      author: response.author,
      license: response.license,
      homePage: response.home_page,
      projectUrls: response.project_urls,
      releases: response.releases,
    };
  }

  // Create virtual environment
  async createVenv(path: string, pythonVersion?: string): Promise<{
    success: boolean;
    path: string;
    pythonVersion: string;
  }> {
    const response = await apiClient.post<{
      success: boolean;
      path: string;
      python_version: string;
    }>("/api/pip/venv/create", {
      path,
      python_version: pythonVersion,
    });

    return {
      success: response.success,
      path: response.path,
      pythonVersion: response.python_version,
    };
  }

  // List virtual environments
  async listVenvs(): Promise<
    Array<{
      name: string;
      path: string;
      pythonVersion: string;
      packagesCount: number;
    }>
  > {
    const response = await apiClient.get<
      Array<{
        name: string;
        path: string;
        python_version: string;
        packages_count: number;
      }>
    >("/api/pip/venv/list");

    return response.map((v) => ({
      name: v.name,
      path: v.path,
      pythonVersion: v.python_version,
      packagesCount: v.packages_count,
    }));
  }

  // Get pip version
  async getVersion(): Promise<string> {
    const response = await apiClient.get<{ version: string }>("/api/pip/version");
    return response.version;
  }

  // Check pip config
  async getConfig(): Promise<Record<string, string>> {
    const response = await apiClient.get<Record<string, string>>("/api/pip/config");
    return response;
  }
}

export const pipService = new PipService();
export default pipService;
